interop_binaries/tests/end_to_end.rs

```diff
@@ -0,0 +1,611 @@
+use anyhow::Context;
+use base64::URL_SAFE_NO_PAD;
+use janus_core::{
+    message::{Duration, TaskId},
+    time::{Clock, RealClock},
+};
+use janus_server::task::PRIO3_AES128_VERIFY_KEY_LENGTH;
+use lazy_static::lazy_static;
+use portpicker::pick_unused_port;
+use prio::codec::Encode;
+use reqwest::{header::CONTENT_TYPE, StatusCode};
+use serde_json::{json, Value};
+use std::{
+    collections::BTreeSet,
+    env, io,
+    net::{Ipv4Addr, SocketAddr},
+    process::{Child, Command, Stdio},
+    time::Duration as StdDuration,
+};
+use testcontainers::{images::postgres::Postgres, RunnableImage};
+use tokio::{
+    io::{AsyncBufReadExt, BufReader},
+    net::TcpStream,
+    process::{ChildStderr, ChildStdout},
+    time::sleep,
+};
+
+static JSON_MEDIA_TYPE: &str = "application/json";
+static MIN_BATCH_DURATION: u64 = 3600;
+
+lazy_static! {
+    static ref CONTAINER_CLIENT: testcontainers::clients::Cli =
+        testcontainers::clients::Cli::default();
+}
+
+/// Wait for a TCP server to begin listening on the given port.
+async fn wait_for_tcp_server(port: u16) -> anyhow::Result<()> {
+    for _ in 0..100 {
+        if TcpStream::connect(SocketAddr::from((Ipv4Addr::LOCALHOST, port)))
+            .await
+            .is_ok()
+        {
+            return Ok(());
+        }
+        sleep(StdDuration::from_millis(200)).await;
+    }
+    Err(anyhow::anyhow!(
+        "Timed out waiting for a server to accept on port {}",
+        port,
+    ))
+}
+
+/// RAII guard to ensure that child processes are cleaned up during test failures.
+struct DropGuard(Vec<Child>);
+
+impl Drop for DropGuard {
+    fn drop(&mut self) {
+        let mut any_error = false;
+        for mut process in self.0.drain(..) {
```

Comment:
This type is complicated by having to deal with multiple `Child`ren IMO. It would be more easily stated if each `DropGuard` owned a single `Child` and implemented the desired drop logic for it.